<div class="alert alert-success">
    
**Комментарий ревьюера**

Привет. Этот проект тоже проверяю я -Сысоева Вера.
    
Давай перейдем к финальной точке - проекту по SQl
</div>

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div class="alert alert-success">
    
**Комментарий ревьюера**

Импорт библиотек и подключение к БД лучше делать в разных ячейках кода, это больше соответствует PEP8.
</div>

# Цели:

- Проанализировать базу данных

- Помочь стартаперы развивать приложения для чтения книг

# Данные о книгах

In [4]:
query = '''SELECT * FROM books'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)


book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

**Талбица books содержит 1000 строк и 6 столбцов**

**Вней хранится информация о книгах:**

- book_id — идентификатор книги;

- author_id — идентификатор автора;

- title — название книги;

- num_pages — количество страниц;

- publication_date — дата публикации книги;

- publisher_id — идентификатор издателя

# Данные об авторах

In [5]:
query = '''SELECT * 
           FROM authors 
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

**Всего в таблице 636 автора чьи имена хранятся в столбце author**

# Данные о пользовательских оценках книг

In [6]:
query = '''SELECT * 
           FROM ratings 
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

**Таблица ratings состоит из 6456 строк и 4 стоблца:**

- rating_id — идентификатор оценки;

- book_id — идентификатор книги;

- username — имя пользователя, оставившего оценку;

- rating — оценка книги

# Данные о пользовательских обзорах

In [7]:
query = '''SELECT * 
           FROM reviews 
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

**Таблица reviews содержит информацию состоящую из следующих столбцов:**


- review_id — идентификатор обзора;

- book_id — идентификатор книги;

- username — имя автора обзора;

- text — текст обзора


# Данные об издательствах

In [8]:
query = '''SELECT * 
           FROM publishers 
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

**Всего в таблице два столбца с publisher_id — идентификатор издательства и publisher — название издательства**

<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**

Ты верно посмотрел содержимое таблиц. Но здесь дополнительно можно было провести стандартный анализ, проверить пропуски, дубликаты и.т.п.
</div>

# Книги вышедшие после 01.01.2000

In [9]:
query = '''SELECT count(book_id )
           FROM books 
           WHERE publication_date > '2000-01-01'
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

**В таблице books была информация о 1000 книгах , и как мы видим из 1000 книг 819 были выпущенны после 01.01.2000**

<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**

Верно
</div>

# Для каждой книги посчитайте количество обзоров и среднюю оценку

In [10]:
query = '''SELECT b.title ,b.book_id ,
           COUNT(distinct r.review_id) as count_review, 
           ROUND(AVG(rat.rating) ,2 )as avg_rating
           FROM books as b
           JOIN reviews as r on b.book_id=r.book_id
           JOIN ratings as rat on r.book_id=rat.book_id
           GROUP BY  b.title , b.book_id
           ORDER BY  COUNT(distinct r.review_id) desc
           
           
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

title  book_id  count_review  avg_rating
0                Twilight (Twilight  #1)      948             7        3.66
1                               The Road      854             6        3.77
2                         The Book Thief      656             6        4.26
3                       The Glass Castle      734             6        4.21
4                    Water for Elephants      963             6        3.98
..                                   ...      ...           ...         ...
989    Naked Empire (Sword of Truth  #8)      465             1        3.50
990                  Moo  Baa  La La La!      446             1        3.00
991  Merrick (The Vampire Chronicles #7)      431             1        4.00
992                            Babyville       92             1        3.50
993                           Lysistrata      413             1        4.00

[994 rows x 4 columns]

In [11]:
query = '''SELECT b.title ,b.book_id, count(distinct r.review_id) as count_review, avg(rat.rating) as avg_rating
           FROM books as b
           left join reviews as r on b.book_id=r.book_id
           left join ratings as rat on r.book_id=rat.book_id
           group by b.title ,b.book_id
           having avg(rat.rating) is null
           order by count(distinct r.review_id) desc
           
           
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

title  book_id  count_review  \
0    Anne Rice's The Vampire Lestat: A Graphic Novel       83             0   
1  Disney's Beauty and the Beast (A Little Golden...      191             0   
2                          Essential Tales and Poems      221             0   
3                               Leonardo's Notebooks      387             0   
4   The Cat in the Hat and Other Dr. Seuss Favorites      672             0   
5                            The Natural Way to Draw      808             0   

  avg_rating  
0       None  
1       None  
2       None  
3       None  
4       None  
5       None

**43 книги имеют оценку 5 , но при это имеют не больше 4 обзора. Больше всего обзоров на книгу Twilight (Twilight #1) со средней оценкой 3.66. Стоит также отметить , что 6 книг не получили обзоров и оценок**

<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**

Хорошо, но стоило одну общую таблицу сделать, включив в том числе и книги без обзоров.
</div>

# Определите издательство, которое выпустило наибольшее число книг толще 50 страниц 

In [12]:
query = ''' SELECT p.publisher,COUNT(b.book_id) 
            FROM books as b
            JOIN publishers as p  on b.publisher_id=p.publisher_id
            WHERE b.num_pages > 50
            GROUP BY   p.publisher
            ORDER BY count(b.book_id) desc
            LIMIT 1
           
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  count
0  Penguin Books     42

**Больше всего книг толщиной более 50 страниц выпустило издательство Penguin Books. Таких книг было 42**

<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**

Да, издатель получен верно
</div>

# Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [15]:
query = '''WITH
            r AS ( SELECT rat.book_id
                    FROM ratings AS rat
                    GROUP BY rat.book_id
                    HAVING COUNT(rating_id)>=50)
            SELECT DISTINCT(a.author), ROUND(AVG(ra.rating),2) AS avg_rating
            FROM authors AS a
            JOIN books AS b on a.author_id=b.author_id
            JOIN r on b.book_id=r.book_id
            JOIN ratings as ra on r.book_id=ra.book_id
            GROUP BY DISTINCT(a.author)
            ORDER BY ROUND(AVG(ra.rating),2) desc
            LIMIT 1
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

**J.K. Rowling/Mary GrandPré вреди всех авторов имеет наивысший средний рейтинг 4.29**

<div class="alert alert-danger">
    
**Комментарий ревьюера**

**#ошибка**

Не совсем верно. Здесь нужно более комплексный запрос составить. Сначала в подзапросе нужно получить только книги с 50 и более оценками, потом объединить этот список с авторами и посчитать для каждого автора среднюю оценку. Сейчас у тебя having работает после groupby, поэтому не верный резульат  получается, одним запросом тут не справиться.
</div>

# Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [14]:
query = ''' 
SELECT AVG(review_count) AS avg_reviews
FROM (
    SELECT r.username, COUNT(r.review_id) AS review_count
    FROM reviews r
    WHERE r.username IN (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 48
    )
    GROUP BY r.username
) AS user_reviews;           '''

con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg_reviews
0         24.0

**Пользователи которые поставили более 48 оценок , в среднем делают 24 обзора**

<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**

Правильно
</div>

# Общий вывод:

- Всего было дано 1000 наименований книг

- 819 книг было выпущено после 01.01.2000

- Есть книги которные не получили ни одного отзыва и оценки 

- Twilight Twilight #1) - книга с самым большим количеством обзоров - 7

- 43 книги имеют наивысшую оценку - 5 

- Penguin Books	- выпустило больше всех книг , у которых более 50 страниц -  42

- J.K. Rowling/Mary GrandPré автор с самой высокой оценкой - 4.29

- Среднее количество обзоров от пользователей - 24


<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**

Теперь мы знаем многие операции с `SQL` в использовании `python` и это отлично, потому что мы будем достаточно часто использовать в нашей практике.
    
Полезные ссылки по SQL:
    
https://habr.com/ru/post/268983/
    
https://habr.com/ru/post/340460/

</div>

<div class="alert alert-success">
    
**Комментарий ревьюера - 2**
    
**#ок**

Все отлично доработано, молодец.
</div>